### PR TITLE
Fixed for JIRA MDM-372

### PR DESCRIPTION
**Pull request by milanperera**

### CODE REVIEW

File: components/policy-mgt/org.wso2.carbon.policy.mgt.core/src/main/java/org/wso2/carbon/policy/mgt/core/dao/FeatureDAO.java [hunk 1]
```diff
@@ -53,4 +53,6 @@
 
     boolean deleteFeaturesOfProfile(Profile profile) throws FeatureManagerDAOException;
 
+    boolean deleteFeaturesOfProfile(int profileId) throws FeatureManagerDAOException;
```
Comment (prabathabey) [inline components/policy-mgt/org.wso2.carbon.policy.mgt.core/src/main/java/org/wso2/carbon/policy/mgt/core/dao/FeatureDAO.java:56]: Add appropriate method level comments.

File: components/policy-mgt/org.wso2.carbon.policy.mgt.core/src/main/java/org/wso2/carbon/policy/mgt/core/dao/ProfileDAO.java [hunk 2]
```diff
@@ -32,6 +32,8 @@
 
     boolean deleteProfile(Profile profile) throws ProfileManagerDAOException;
 
+    boolean deleteProfile(int policyId) throws ProfileManagerDAOException;
+
```
Comment (prabathabey) [inline components/policy-mgt/org.wso2.carbon.policy.mgt.core/src/main/java/org/wso2/carbon/policy/mgt/core/dao/ProfileDAO.java:36]: Add appropriate method level comments.

File: components/policy-mgt/org.wso2.carbon.policy.mgt.core/src/main/java/org/wso2/carbon/policy/mgt/core/dao/impl/FeatureDAOImpl.java [hunk 3]
```diff
@@ -261,6 +261,28 @@ public boolean deleteFeaturesOfProfile(Profile profile) throws FeatureManagerDAO
...
+            conn = this.getConnection();
+            String query = "DELETE FROM DM_PROFILE_FEATURES WHERE PROFILE_ID = ?";
+            stmt = conn.prepareStatement(query);
+            stmt.setInt(1, profileId);
+            stmt.executeUpdate();
+            return true;
+
+        } catch (SQLException e) {
+            String msg = "Error occurred while deleting the feature related to a profile.";
+            log.error(msg);
```
Comment (prabathabey) [inline components/policy-mgt/org.wso2.carbon.policy.mgt.core/src/main/java/org/wso2/carbon/policy/mgt/core/dao/impl/FeatureDAOImpl.java:279]: Avoid log and throw unless specifically needed

File: components/policy-mgt/org.wso2.carbon.policy.mgt.core/src/main/java/org/wso2/carbon/policy/mgt/core/dao/impl/ProfileDAOImpl.java [hunk 4]
```diff
@@ -155,6 +155,29 @@ public boolean deleteProfile(Profile profile) throws ProfileManagerDAOException
...
+            conn = this.getConnection();
+            String query = "DELETE FROM DM_PROFILE WHERE ID = ?";
+            stmt = conn.prepareStatement(query);
+            stmt.setInt(1, profileId);
+            stmt.executeUpdate();
+            return true;
+
+        } catch (SQLException e) {
+            String msg = "Error occurred while deleting the profile from the data base.";
+            log.error(msg);
```
Comment (prabathabey) [inline components/policy-mgt/org.wso2.carbon.policy.mgt.core/src/main/java/org/wso2/carbon/policy/mgt/core/dao/impl/ProfileDAOImpl.java:172]: Avoid log and throw unless specifically needed

File: components/policy-mgt/org.wso2.carbon.policy.mgt.core/src/main/java/org/wso2/carbon/policy/mgt/core/dao/impl/FeatureDAOImpl.java [hunk 5]
```diff
@@ -261,6 +261,28 @@ public boolean deleteFeaturesOfProfile(Profile profile) throws FeatureManagerDAO
         }
     }
 
+    @Override
+    public boolean deleteFeaturesOfProfile(int profileId) throws FeatureManagerDAOException {
+        Connection conn;
+        PreparedStatement stmt = null;
```
Comment (DilanUA) [inline components/policy-mgt/org.wso2.carbon.policy.mgt.core/src/main/java/org/wso2/carbon/policy/mgt/core/dao/impl/FeatureDAOImpl.java:267]: unnecessary null assignment here...

File: components/policy-mgt/org.wso2.carbon.policy.mgt.core/src/main/java/org/wso2/carbon/policy/mgt/core/dao/impl/FeatureDAOImpl.java [hunk 6]
```diff
@@ -261,6 +261,28 @@ public boolean deleteFeaturesOfProfile(Profile profile) throws FeatureManagerDAO
...
+
+        try {
+            conn = this.getConnection();
+            String query = "DELETE FROM DM_PROFILE_FEATURES WHERE PROFILE_ID = ?";
+            stmt = conn.prepareStatement(query);
+            stmt.setInt(1, profileId);
+            stmt.executeUpdate();
+            return true;
+
```
Comment (DilanUA) [inline components/policy-mgt/org.wso2.carbon.policy.mgt.core/src/main/java/org/wso2/carbon/policy/mgt/core/dao/impl/FeatureDAOImpl.java:276]: unnecessary empty line here here...

File: components/policy-mgt/org.wso2.carbon.policy.mgt.core/src/main/java/org/wso2/carbon/policy/mgt/core/dao/impl/ProfileDAOImpl.java [hunk 7]
```diff
@@ -155,6 +155,29 @@ public boolean deleteProfile(Profile profile) throws ProfileManagerDAOException
...
+        PreparedStatement stmt = null;
+
+        try {
+            conn = this.getConnection();
+            String query = "DELETE FROM DM_PROFILE WHERE ID = ?";
+            stmt = conn.prepareStatement(query);
+            stmt.setInt(1, profileId);
+            stmt.executeUpdate();
+            return true;
+
```
Comment (DilanUA) [inline components/policy-mgt/org.wso2.carbon.policy.mgt.core/src/main/java/org/wso2/carbon/policy/mgt/core/dao/impl/ProfileDAOImpl.java:169]: unnecessary empty line here here...

File: components/policy-mgt/org.wso2.carbon.policy.mgt.core/src/main/java/org/wso2/carbon/policy/mgt/core/dao/impl/FeatureDAOImpl.java [hunk 8]
```diff
@@ -261,6 +261,28 @@ public boolean deleteFeaturesOfProfile(Profile profile) throws FeatureManagerDAO
         }
     }
 
+    @Override
+    public boolean deleteFeaturesOfProfile(int profileId) throws FeatureManagerDAOException {
```
Comment (DilanUA) [inline components/policy-mgt/org.wso2.carbon.policy.mgt.core/src/main/java/org/wso2/carbon/policy/mgt/core/dao/impl/FeatureDAOImpl.java:265]: deleteFeaturesOfProfile does not sound like returning a boolean. BTW, do we really need to return true here if transaction is successfull?

File: components/policy-mgt/org.wso2.carbon.policy.mgt.core/src/main/java/org/wso2/carbon/policy/mgt/core/dao/impl/ProfileDAOImpl.java [hunk 9]
```diff
@@ -155,6 +155,29 @@ public boolean deleteProfile(Profile profile) throws ProfileManagerDAOException
     }
 
     @Override
+    public boolean deleteProfile(int profileId) throws ProfileManagerDAOException {
```
Comment (DilanUA) [inline components/policy-mgt/org.wso2.carbon.policy.mgt.core/src/main/java/org/wso2/carbon/policy/mgt/core/dao/impl/ProfileDAOImpl.java:158]: deleteProfile does not sound like returning a boolean. BTW, do we really need to return true here if transaction is successfull?

File: components/policy-mgt/org.wso2.carbon.policy.mgt.core/src/main/java/org/wso2/carbon/policy/mgt/core/mgt/impl/PolicyManagerImpl.java [hunk 10]
```diff
@@ -236,12 +236,14 @@ public boolean updatePolicyPriorities(List<Policy> policies) throws PolicyManage
...
         try {
             PolicyManagementDAOFactory.beginTransaction();
-            bool = policyDAO.deletePolicy(policy);
+            policyDAO.deleteAllPolicyRelatedConfigs(policy.getId());
+            policyDAO.deletePolicy(policy.getId());
+            featureDAO.deleteFeaturesOfProfile(policy.getProfileId());
+            profileDAO.deleteProfile(policy.getProfileId());
             PolicyManagementDAOFactory.commitTransaction();
-
+            return true;
```
Comment (DilanUA) [inline components/policy-mgt/org.wso2.carbon.policy.mgt.core/src/main/java/org/wso2/carbon/policy/mgt/core/mgt/impl/PolicyManagerImpl.java:246]: deletePolicy does not sound like returning a boolean. BTW, do we really need to return true here if transaction is successfull?

File: components/policy-mgt/org.wso2.carbon.policy.mgt.core/src/main/java/org/wso2/carbon/policy/mgt/core/mgt/impl/PolicyManagerImpl.java [hunk 11]
```diff
@@ -249,21 +251,51 @@ public boolean deletePolicy(Policy policy) throws PolicyManagementException {
...
+
+            if (log.isDebugEnabled()) {
+                log.debug("Profile ID: " + policy.getProfileId());
+            }
+
+            featureDAO.deleteFeaturesOfProfile(policy.getProfileId());
+
+            profileDAO.deleteProfile(policy.getProfileId());
             PolicyManagementDAOFactory.commitTransaction();
+            return true;
```
Comment (harshanL) [inline components/policy-mgt/org.wso2.carbon.policy.mgt.core/src/main/java/org/wso2/carbon/policy/mgt/core/mgt/impl/PolicyManagerImpl.java:298]: This method always returns true. So can't we simply have void or change the logic to return a bool value depending on the status?

File: components/policy-mgt/org.wso2.carbon.policy.mgt.core/src/main/java/org/wso2/carbon/policy/mgt/core/dao/impl/FeatureDAOImpl.java (hunk 8 above)
Comment (milanperera) [inline components/policy-mgt/org.wso2.carbon.policy.mgt.core/src/main/java/org/wso2/carbon/policy/mgt/core/dao/impl/FeatureDAOImpl.java:265]: This boolean is used just to acknowledge whether the operation is successful. It returns true if there are any affected rows of the query.

File: components/policy-mgt/org.wso2.carbon.policy.mgt.core/src/main/java/org/wso2/carbon/policy/mgt/core/dao/impl/FeatureDAOImpl.java (hunk 5 above)
Comment (milanperera) [inline components/policy-mgt/org.wso2.carbon.policy.mgt.core/src/main/java/org/wso2/carbon/policy/mgt/core/dao/impl/FeatureDAOImpl.java:267]: The 'stmt' variable should be initialized prior to passing to 'cleanupResources(stmt, resultset)' method

File: components/policy-mgt/org.wso2.carbon.policy.mgt.core/src/main/java/org/wso2/carbon/policy/mgt/core/dao/impl/ProfileDAOImpl.java (hunk 9 above)
Comment (milanperera) [inline components/policy-mgt/org.wso2.carbon.policy.mgt.core/src/main/java/org/wso2/carbon/policy/mgt/core/dao/impl/ProfileDAOImpl.java:158]: This boolean is used just to acknowledge whether the operation is successful. It returns true if there are any affected rows of the query.

File: components/policy-mgt/org.wso2.carbon.policy.mgt.core/src/main/java/org/wso2/carbon/policy/mgt/core/mgt/impl/PolicyManagerImpl.java (hunk 10 above)
Comment (milanperera) [inline components/policy-mgt/org.wso2.carbon.policy.mgt.core/src/main/java/org/wso2/carbon/policy/mgt/core/mgt/impl/PolicyManagerImpl.java:246]: This boolean is used just to acknowledge whether the operation is successful. It returns true if there are any affected rows of the query.

File: components/policy-mgt/org.wso2.carbon.policy.mgt.core/src/main/java/org/wso2/carbon/policy/mgt/core/dao/impl/FeatureDAOImpl.java (hunk 3 above)
Comment (milanperera) [inline components/policy-mgt/org.wso2.carbon.policy.mgt.core/src/main/java/org/wso2/carbon/policy/mgt/core/dao/impl/FeatureDAOImpl.java:279]: Fixed in commit [1].
[1] https://github.com/wso2/carbon-device-mgt/commit/cdbdc86f42b350c24e9d059ff7c23e1f09242df4

File: components/policy-mgt/org.wso2.carbon.policy.mgt.core/src/main/java/org/wso2/carbon/policy/mgt/core/dao/impl/ProfileDAOImpl.java (hunk 4 above)
Comment (milanperera) [inline components/policy-mgt/org.wso2.carbon.policy.mgt.core/src/main/java/org/wso2/carbon/policy/mgt/core/dao/impl/ProfileDAOImpl.java:172]: Fixed in commit [1].
[1] https://github.com/wso2/carbon-device-mgt/commit/cdbdc86f42b350c24e9d059ff7c23e1f09242df4

File: components/policy-mgt/org.wso2.carbon.policy.mgt.core/src/main/java/org/wso2/carbon/policy/mgt/core/dao/FeatureDAO.java (hunk 1 above)
Comment (milanperera) [inline components/policy-mgt/org.wso2.carbon.policy.mgt.core/src/main/java/org/wso2/carbon/policy/mgt/core/dao/FeatureDAO.java:56]: Added in PR [1].
[1] https://github.com/wso2/carbon-device-mgt/pull/27

File: components/policy-mgt/org.wso2.carbon.policy.mgt.core/src/main/java/org/wso2/carbon/policy/mgt/core/dao/ProfileDAO.java (hunk 2 above)
Comment (milanperera) [inline components/policy-mgt/org.wso2.carbon.policy.mgt.core/src/main/java/org/wso2/carbon/policy/mgt/core/dao/ProfileDAO.java:36]: Added in PR [1].
[1] https://github.com/wso2/carbon-device-mgt/pull/27

File: components/policy-mgt/org.wso2.carbon.policy.mgt.core/src/main/java/org/wso2/carbon/policy/mgt/core/dao/impl/FeatureDAOImpl.java (hunk 6 above)
Comment (milanperera) [inline components/policy-mgt/org.wso2.carbon.policy.mgt.core/src/main/java/org/wso2/carbon/policy/mgt/core/dao/impl/FeatureDAOImpl.java:276]: Removed in PR [1].
[1] https://github.com/wso2/carbon-device-mgt/pull/27

File: components/policy-mgt/org.wso2.carbon.policy.mgt.core/src/main/java/org/wso2/carbon/policy/mgt/core/dao/impl/ProfileDAOImpl.java (hunk 7 above)
Comment (milanperera) [inline components/policy-mgt/org.wso2.carbon.policy.mgt.core/src/main/java/org/wso2/carbon/policy/mgt/core/dao/impl/ProfileDAOImpl.java:169]: Removed in PR [1].
[1] https://github.com/wso2/carbon-device-mgt/pull/27

File: components/policy-mgt/org.wso2.carbon.policy.mgt.core/src/main/java/org/wso2/carbon/policy/mgt/core/mgt/impl/PolicyManagerImpl.java (hunk 11 above)
Comment (milanperera) [inline components/policy-mgt/org.wso2.carbon.policy.mgt.core/src/main/java/org/wso2/carbon/policy/mgt/core/mgt/impl/PolicyManagerImpl.java:298]: Implemented in PR [1].
[1] https://github.com/wso2/carbon-device-mgt/pull/27

**Pull request merged**